paramiko/hostkeys.py

```diff
@@ -196,14 +197,34 @@ def _hostname_matches(self, hostname, entry):
         """
         for h in entry.hostnames:
             if (
-                h == hostname or
+                self.wildcard_matches(h, hostname) or
                 h.startswith('|1|') and
                 not hostname.startswith('|1|') and
                 constant_time_bytes_eq(self.hash_host(hostname, h), h)
             ):
                 return True
         return False
 
+    def wildcard_matches(self, wildcard, hostname):
+        """
+        Implements '*' and '?' wildcards and '!' negation as in sshd.
+        :param str wildcard: host name / pattern from known_hosts
+        :param str hostname: target host name
+        :return bool:
+        """
+        wildcard_negation = False
+        if hostname.startswith('|1|'):
+            return False
+        if wildcard.startswith("!"):
+            wildcard_negation = True
+
+        wildcard = wildcard.lstrip("!")
+        wildcard = re.escape(wildcard)
+        wildcard = wildcard.replace("\\?", ".?")
+        wildcard = wildcard.replace("\\*", ".*")
+
+        return re.match(wildcard, hostname) and wildcard_negation is not True
```

Comment:
It is simpler to write `and not` instead of `is not True`. But we see here that this can't really support negation properly - it just disables the pattern completely. You might as well `return False` above for `if wildcard.startswith("!")`. To support negation properly you would have to make a match here cause all other hostnames/patterns on the same line to also fail to match. For that, you would need a whole different way to parse and handle `known_hosts` in paramiko.